### PR TITLE
ArC - fix ordering of multiple decorators with the same priority

### LINKED ISSUE #18538: NPE during build when multiple decorators have the same priority
## Describe the bug
NPE during build when multiple decorators have the same priority (or no priority defined)

### Expected behavior
App builds and both decorators are instantiated properly (according to CDI spec execution order is undefined for decorators with the same priority though)

### Actual behavior
```
[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:2.0.1.Final:build (default) on project app: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR] 	[error]: Build step io.quarkus.arc.deployment.ArcProcessor#generateResources threw an exception: java.lang.NullPointerException
[ERROR] 	at io.quarkus.gizmo.BytecodeCreatorImpl.allocateLocalVariables(BytecodeCreatorImpl.java:1161)
[ERROR] 	at io.quarkus.gizmo.MethodCreatorImpl.write(MethodCreatorImpl.java:102)
[ERROR] 	at io.quarkus.gizmo.ClassCreator.writeTo(ClassCreator.java:180)
[ERROR] 	at io.quarkus.gizmo.ClassCreator.close(ClassCreator.java:203)
[ERROR] 	at io.quarkus.arc.processor.SubclassGenerator.generate(SubclassGenerator.java:122)
[ERROR] 	at io.quarkus.arc.processor.BeanProcessor.generateResources(BeanProcessor.java:200)
[ERROR] 	at io.quarkus.arc.deployment.ArcProcessor.generateResources(ArcProcessor.java:457)
[ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[ERROR] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR] 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
[ERROR] 	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:820)
[ERROR] 	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
[ERROR] 	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2442)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1476)
[ERROR] 	at java.base/java.lang.Thread.run(Thread.java:829)
[ERROR] 	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
```

## To Reproduce
[main-decorator.zip](https://github.com/quarkusio/quarkus/files/6786572/main-decorator.zip)

### Output of `uname -a` or `ver`
`Darwin xtaixe 20.5.0 Darwin Kernel Version 20.5.0: Sat May  8 05:10:33 PDT 2021; root:xnu-7195.121.3~9/RELEASE_X86_64 x86_64`

### Output of `java -version`
```
openjdk version "11.0.11" 2021-04-20
OpenJDK Runtime Environment GraalVM CE 21.1.0 (build 11.0.11+8-jvmci-21.1-b05)
OpenJDK 64-Bit Server VM GraalVM CE 21.1.0 (build 11.0.11+8-jvmci-21.1-b05, mixed mode, sharing)
```

### Quarkus version or git rev
`2.0.1.Final`

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)
```
Apache Maven 3.8.1 (05c21c65bdfed0f71a2f2ada8b84da59348c4c5d)
Maven home: /Users/xtaixe/.mvnvm/apache-maven-3.8.1
Java version: 11.0.11, vendor: GraalVM Community, runtime: /Library/Java/JavaVirtualMachines/graalvm-ce-java11-21.1.0/Contents/Home
Default locale: en_US, platform encoding: UTF-8
OS name: "mac os x", version: "11.4", arch: "x86_64", family: "mac"
```

### PR DESCRIPTION
**Pull request by mkouba**
- resolves #18538

### CODE REVIEW

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/SubclassGenerator.java [hunk 1]
```diff
@@ -421,7 +421,8 @@ private void processDecorator(DecoratorInfo decorator, BeanInfo bean, Type provi
         ClassCreator delegateSubclass = delegateSubclassBuilder.build();
 
         Map<MethodDescriptor, DecoratorInfo> nextDecorators = bean.getNextDecorators(decorator);
-        List<DecoratorInfo> decoratorParameters = new ArrayList<>(nextDecorators.values());
+        Collection<DecoratorInfo> nextDecoratorsValues = nextDecorators.values();
+        List<DecoratorInfo> decoratorParameters = new ArrayList<>(new HashSet<>(nextDecoratorsValues));
```
Comment (mkouba) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/SubclassGenerator.java:425]: This is actually a different fix, or a small optimization - we don't need to create a "next decorator" field for each decorated method if the same decorator is used. The same reference would be set anyway.

**Review by manovotn — APPROVED** [review]
Looks good.

**Comment by mkouba** [comment]
> `org.eclipse.microprofile.fault.tolerance.tck.TimeoutUninterruptableTest.testTimeoutAsyncBulkhead` line `190` - [More details](https://github.com/quarkusio/quarkus/runs/3029014837#user-content-test-failure-org.eclipse.microprofile.fault.tolerance.tck.timeoutuninterruptabletest-1) - [Source on GitHub](https://github.com/mkouba/quarkus/blob/c7cc3c11c75d066de4237fa21c2fc252a7a2791d/tcks/microprofile-fault-tolerance/src/test/java/org/eclipse/microprofile/fault/tolerance/tck/TimeoutUninterruptableTest.java#L190)

@Ladicek Any idea?

**Comment by famod** [comment]
That test has been flaky for quite some time now: https://github.com/quarkusio/quarkus/pulls?q=is%3Apr+testTimeoutAsyncBulkhead

**Pull request merged**